### ISSUE TITLE
Doubled Escaping on Generated PHP code

### ISSUE THREAD

**Issue by exafred** [firasdib/Regex101#526]
A PHP regex generated using the Code Generator had been doubly-escaped.

`^([a-z]{1,2}[a-z0-9]{1,2})\s*?(\d[a-z]{2})$` became `^([a-z]{1,2}[a-z0-9]{1,2})\\s*?(\\d[a-z]{2})$` in the generated code.

This didn't apply to either Python or JS versions of generated code.

**Comment by OnlineCop** [comment]
The TL;DR of this is: This is correct behavior, and by design.

Javascript doesn't use a string wrapped in double quotes, where PHP does. Since you're writing to a string, the `\\` in `\\s`  becomes simply `\` (thereby leaving the result as `\s`  that PHP actually sees). The other languages aren't interpreting the backslash character into a string (in the case of Python, the use of single quotes keeps a backslash as a literal backslash without interpreting it as a meta character that will make a `\n` into a newline, for example).

I've also used the PHP output for languages like C# or Java that also require all the `\` characters to be double-escaped when they're within strings.

**Comment by nhahtdh** [comment]
@OnlineCop Rather than because escape is necessary, it's because it's easier to generate code by escaping all of them. In PHP, escape sequences that are not recognized retains both the backslash and the following character. I guess we can leave this open as an enhancement.

**Comment by exafred** [comment]
Well, I certainly stand educated! Thanks both for explaining and elaborating upon this.

**Comment by nhahtdh** [comment]
@OnlineCop I think you confused Python and PHP's string literal:

- PHP's double quoted string does string interpolation and recognize a number of escape sequence, while its single quoted string doesn't do string interpolation, and only recognize `\'` and `\\` escape sequences - unrecognized escape sequence retains the backslash `\`.
- Python has "normal" string and raw string, differentiated by prefix `r` on the string. The quotes to surround the string (single/double quote) are interchangeable. Similar to PHP, raw string only recognize `\\` and (`\"` or `\'` - depending on the quote), but all characters in the sequence are retained.

**Comment by firasdib** [comment]
Is this still relevant on v2?

**Comment by nhahtdh** [comment]
I need some time to think over this issue. It's probably fixed, but I'm too tired now to think clearly.

**Comment by firasdib** [comment]
In version2 this becomes

```php
$re = '/^([a-z]{1,2}[a-z0-9]{1,2})\s*?(\d[a-z]{2})$/';
```

Which I think is correct.

**Comment by firasdib** [comment]
@nhahtdh If this is fixed please close

**Comment by nhahtdh** [comment]
There seems to be an error with this test case:

- Regex: `/\\'\^/`
- Input: `\'^`

The generated code is:
```
$re = '/\\\'\^/';
$str = '\\\'^';
```

which [outputs the wrong result](https://ideone.com/O5hjbC):
```
Array
(
    [0] => Array
        (
            [0] => '^
        )

)
```

It should be:

```
$re = '/\\\\\'\^/';
$str = '\\\'^';
```

There was this [duplicated issue]( https://github.com/firasdib/Regex101/issues/41) in the past for version 1, when you used single quote string literal for regex before switching to double quote string literal.

**Comment by firasdib** [comment]
Fixed in v2,  please verify and close.

**Comment by nhahtdh** [comment]
@firasdib I have verified, and while the behavior is correct, it doesn't handle this enhancement. This enhancement is meant to avoid escaping `\` where unnecessary in the regex string.

This is the algorithm for minimal escaping, assuming we use single quoted string:
```
# The code below in a loop
# curr is the index of the current position
switch re_str[curr]
  case \
    if re_str[curr + 1] == \
        append 4 backslash
    else
        append backslash + re_str[curr+1]
    curr++
    break
  case '
    append backslash + single quote
    break
  case delimiter
    append backslash + re_str[curr]
   break
  default
    append re_str[curr]
curr++
```

**Comment by firasdib** [comment]
@nhahtdh If I were to switch to double quotes this issue would not exist, correct?

**Comment by nhahtdh** [comment]
It still exists, and becomes worse.

**Comment by firasdib** [comment]
I've implemented your algorithm and pushed it live. Please verify and close.

**Comment by nhahtdh** [comment]
Confirm fixed :+1:
